modules/navier_stokes/src/kernels/INSMass.C

```diff
@@ -17,35 +17,77 @@ InputParameters validParams<INSMass>()
   params.addCoupledVar("w", 0, "z-velocity"); // only required in 3D
   params.addRequiredCoupledVar("p", "pressure");
 
+  MooseEnum coord_types("XYZ RZ RSPHERICAL");
+  params.addParam<MooseEnum>("coord_type", coord_types, "Coordinate system types. Choices are: " + coord_types.getRawNames());
   return params;
 }
 
 
 
 INSMass::INSMass(const InputParameters & parameters) :
-  Kernel(parameters),
-
-  // Gradients
-  _grad_u_vel(coupledGradient("u")),
-  _grad_v_vel(coupledGradient("v")),
-  _grad_w_vel(coupledGradient("w")),
-
-  // Variable numberings
-  _u_vel_var_number(coupled("u")),
-  _v_vel_var_number(coupled("v")),
-  _w_vel_var_number(coupled("w")),
-  _p_var_number(coupled("p"))
+    Kernel(parameters),
+
+    _coord_type_set(false),
+
+    // Velocities (important for non-cartesian coordinates)
+    _u_vel(coupledValue("u")),
+    _v_vel(coupledValue("v")),
+    _w_vel(coupledValue("w")),
+
+    // Gradients
+    _grad_u_vel(coupledGradient("u")),
+    _grad_v_vel(coupledGradient("v")),
+    _grad_w_vel(coupledGradient("w")),
+
+    // Variable numberings
+    _u_vel_var_number(coupled("u")),
+    _v_vel_var_number(coupled("v")),
+    _w_vel_var_number(coupled("w")),
+    _p_var_number(coupled("p"))
+
+{
+}
+
+void INSMass::setGeometryParameter(const InputParameters & params,
+                                   INSMass::COORD_TYPE & coord_type)
 {
+  if (params.isParamSetByUser("coord_type"))
+  {
+    coord_type = INSMass::COORD_TYPE(int(params.get<MooseEnum>("coord_type")));
+  }
+  else
+  {
+    coord_type = INSMass::XYZ;
+  }
 }
 
+void INSMass::computeResidual()
+{
+  if (!_coord_type_set)
+  {
+    setGeometryParameter(_pars, _coord_type);
+    _coord_type_set = true;
+  }
 
+  Kernel::computeResidual();
+}
 
 Real INSMass::computeQpResidual()
 {
   // (div u) * q
   // Note: we (arbitrarily) multilply this term by -1 so that it matches the -p(div v)
   // term in the momentum equation.  Not sure if that is really important?
-  return -(_grad_u_vel[_qp](0) + _grad_v_vel[_qp](1) + _grad_w_vel[_qp](2)) * _test[_i][_qp];
+  switch(_coord_type)
```

Comment:
Civet is currently failing because you need a space between `switch` and `(_coord_type)`